README


This archive contains the distribution Sort-Hash,
version 2.04:

  Sort the keys of a Hash into an Array without having to read the perldoc for sort!

This software is Copyright (c) 2014 by John Karr.

This is free software, licensed under:

  The GNU General Public License, Version 3, June 2007


This README file was generated by Dist::Zilla::Plugin::Readme v5.008.



Changes
Revision history for Sort::Hash

2.04	25 July 2014
	META.yml fix, attempting to prevent indexing of Sort::Hash1.
2.02,203    15 July 2014
        Pod Fixes.
        Remove version number from Sort::Hash1 because Kwalitee didn't like it.
2.01    15 July 2014
	Fixed a minor pod error
2.00    14 July 2014
	Rewrote interface
1.05    12 July 2014
		Added Fatal switch to enable choice of dying or warning.
		When fatal is off a failed sort returns an empty set.
1.04    9 April 2014
        Minor Doc fix to improve CPANTS Kwalitee.
	    1.041 fixed version number in POD.
1.03    8 February 2014
        Fixed error in MakeFile to specify minimum pre-reqs.
1.02    7 February 2014
        attempting to numeric sort an alpha value warns and returns undef.
        added alphastrict to reject alpha sort with numeric values.
1.01    6 February 2014
        Specified minimum Perl Version of 5.8
        Added Changelog.
1.00    4 February 2014
        First version, released on an unsuspecting world.

lib/Sort/Hash.pm
use strict;
use warnings FATAL => 'all';

package Sort::Hash;
{
  $Sort::Hash::VERSION = '2.04';
}
use Exporter 'import';
use Try::Tiny 0.13;
use Scalar::Util 1.24;

our @EXPORT = qw( sort_hash );    # symbols to export on request

# ABSTRACT: Sort the keys of a Hash into an Array without having to read the perldoc for sort!

=pod

=head1 NAME

Sort::Hash - get the keys to a hashref sorted by their values

=head1 VERSION

version 2.04

=head1 SYNOPSIS

Hash::Sort is just a convenience for returning the keys of a hashref
sorted by their values. The hash may either be passed directly or
as a hashref, numeric and alphanumeric sorting are supported, 
the sort may be either Ascending or Descending. 

  use Sort::Hash;
  my @sorted = sort_hash( \%Hash );

This does exactly the same as: 

 my @sorted = ( sort { $Hash{$a} <=> $Hash{$b} } keys %Hash ) ;

=head1 DESCRIPTION 

Sort the keys of a Hash into an Array without having to read the perldoc for sort! 

=head2 sort_hash 

Return a sorted array containing the keys of a hash.

=head3 Options to sort_hash

    nofatal      warn and return an emppty list instead of dying on
                 invalid sort
    silent       like nofatal but doesn't emit warnings either
    noempty      if the hashref is empty treat it as an error
                 instead of returning an empty list ().
    desc         sort descending instead of ascending
    asc          ascending sort is the default but you can specify it
    alpha        sort alpha (treats numbers as text)
    strictalpha  sort alpha but refuse to sort numbers as text
    numeric      sort as numbers, default is numeric

The first argument is the hashref to be sorted, followed by the arguments which may be in any order.

 sort_hash( $hashref, 'strictalpha', 'desc' );
 sort_hash( $hashref, qw/ noempty nofatal alpha desc /);

=head2 Errors

Numeric sorts will fail if given a non-number. Normally alpha sorts will
treat numbers as text. strictalpha uses Scalar::Util::looks_like_number 
to reject a hash that has any values that appear to be numbers.

When the data is illegal for the sort type in effect, (only alpha has no restriction) sort_hash will die. If you prefer it not to, use nofatal to return () and warn instead of die, silent (implies nofatal) will just return () without a warning. 

Sorting an empty hashref will return nothing (). You can make this into an error that will die or warn depending on the nofatal flag with noempty.

=head1 Changes from Version 1.x to 2.x

The API has been changed from version 1. When looking back at sort_hash I decided that I didn't like a number of things about it, I even ended up agreeing with a complaint from PerlCritic! 

=head2 If you need version1 compatibility

Version 1 is included in the version 2 distribution, renamed as Sort::Hash1, just change your use statement to C<use Sort::Hash1;>.

=cut

sub sort_hash {
    my @sorted = ();
    my $H      = shift;
    my ( $silent, $nofatal, $noempty, $desc, $alpha, $strictalpha ) = 0;
    my ( $numeric, $asc ) = 1;
    for (@_) {
        if ( $_ eq 'nofatal' ) { $nofatal = 1 }
        if ( $_ eq 'silent' )  { $silent  = 1; $nofatal = 1 }
        if ( $_ eq 'noempty' ) { $noempty = 1 }
        if ( $_ eq 'desc' )    { $desc    = 1; $asc = 0 }
        if ( $_ eq 'asc' )     { $asc     = 1; $desc = 0 }
        if ( $_ eq 'alpha' )   { $alpha   = 1; $numeric = 0; }
        if ( $_ eq 'strictalpha' ) {
            $strictalpha = 1;
            $alpha       = 1;
            $numeric     = 0;
        }
        if ( $_ eq 'numeric' ) { $strictalpha = 0; $alpha = 0; $numeric = 1; }
    }

    my $death = sub {
        if ($nofatal) { warn $_[0] unless $silent; return (); }
        else          { die $_[0]; }
    };
    if ($noempty) {
        unless ( scalar( keys %$H ) ) {
            $death->(
                'Attempt to sort an empty hash while noempty is in effect');
        }
    }
    if ($strictalpha) {
        for ( values %{$H}) {
            if ( Scalar::Util::looks_like_number($_) ) {
                $death->(
                    'Attempt to Sort Numeric Value in Strict Alpha Sort');
                return ;
            }
        }
    }
    if ($alpha) {
        @sorted = ( sort { lc $H->{$a} cmp lc $H->{$b} } keys %{$H} );
    }
    else {
        try {
            @sorted = ( sort { $H->{$a} <=> $H->{$b} } keys %{$H} );
        }
        catch {
            $death->('Attempt to Sort non-Numeric values in a Numeric Sort');
            return ;
        }
    }
    if ( $desc ) {
        return reverse @sorted;
    }
    else { return @sorted; }
}

=pod

=head1 AUTHOR

John Karr, C<brainbuz at brainbuz.org>

=head1 BUGS

Please report any bugs or feature requests via the BitBucket issue tracker at
L<https://bitbucket.org/brainbuz/sort-hash/issues>.  I will be
notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

You can also look for information at: The documentation for the 
sort command in the Perl documentation.

=head1 LICENSE AND COPYRIGHT

Copyright 2014 John Karr.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 3 or at your option
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree;
if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

=cut

1;


lib/Sort/Hash1.pm
package Sort::Hash1;
{
  $Sort::Hash1::VERSION = '2.04';
}
use Exporter 'import';
use Try::Tiny 0.13;
use Scalar::Util 1.24;
use strict;
use warnings FATAL => 'all';

our @EXPORT = qw( sort_hash );    # symbols to export on request

=head1 NAME

Sort::Hash1 Version1 of SortHash

=head1 VERSION 

1.042

=head1 SYNOPSIS

Sort::Hash1 is included in the distribution for backwards compatibility. If you used Sort::Hash and don't want to fix the code right now, use Sort::Hash1. If you are writing or maintaining your code use Sort::Hash instead. 

=head2 sort_hash

The interface to sort_hash changed from version 1 to version 2. Instead of taking a hash of values it now takes a list in which the first item must be a reference to the hash to sort. Sort::Hash1 will be removed from the distribution in the future, updating your code is easy:

 sort_hash( desc => 1 , alpha => 1, %somehash );
 becomes
 sort_hash( \%somehash, 'desc', 'alpha' );

=cut

sub sort_hash {
    my %H      = @_;
    my @sorted = ();
    my $direction   = delete $H{direction}   || 'asc';
    my $alpha       = delete $H{alpha}       || 0;
    my $strictalpha = delete $H{strictalpha} || 0;
    my $numeric     = delete $H{numeric}     || 1;
    if ( defined $H{hashref} ) { %H = %{ $H{hashref} } }
    if ($strictalpha) { 
        $alpha = 1;
        for ( values %H ) {
            if (Scalar::Util::looks_like_number($_)) {
                warn 'Attempt to Sort Numeric Value in Strict Alpha Sort';
                return undef }
            }
        }
    if ($alpha) {
        @sorted = ( sort { lc $H{$a} cmp lc $H{$b} } keys %H ) ;
        }
    else {
        try { @sorted = ( sort { $H{$a} <=> $H{$b} } keys %H ) ;}
        catch { 
            warn 'Attempt to Sort non-Numeric values in a Numeric Sort';
            return undef ; }
        }
    if ( lc($direction) eq 'desc' ) {
        return reverse @sorted;
    }
    else { return @sorted; }
}

=head1 AUTHOR

John Karr, C<< <brainbuz at brainbuz.org> >>

=head1 BUGS

Please report any bugs or feature requests via the BitBucket issue tracker at
L<https://bitbucket.org/brainbuz/sort-hash/issues>.  I will be
notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

You can also look for information at: The documentation for the 
sort command in the Perl documentation.

=head1 LICENSE AND COPYRIGHT

Copyright 2014 John Karr.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 3 or at your option
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree;
if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

=cut

1;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.008.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Sort the keys of a Hash into an Array without having to read the perldoc for sort!",
  "AUTHOR" => "John Karr <brainbuz\@brainbuz.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Sort-Hash",
  "EXE_FILES" => [],
  "LICENSE" => "gpl",
  "NAME" => "Sort::Hash",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "Scalar::Util" => "1.24",
    "Try::Tiny" => "0.13",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test::Exception" => 0,
    "Test::More" => 0,
    "Test::Warn" => 0,
    "Test::Warnings" => 0
  },
  "VERSION" => "2.04",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => 0,
  "Scalar::Util" => "1.24",
  "Test::Exception" => 0,
  "Test::More" => 0,
  "Test::Warn" => 0,
  "Test::Warnings" => 0,
  "Try::Tiny" => "0.13",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



